### PR TITLE
KEYCLOAK-7969 - SAML users should not be identified by SAML:NameID

### PR DESCRIPTION
**Pull request by dteleguin**
This PR adds an option to configure principal identifier for a SAML IdP, which can be either Subject NameID (default), or a SAML attribute (either by name or friendly name).

The principal value will then go to the FederatedIdentity's `userId` and `userName` fields, and will be used to identify and track external identities.

### CODE REVIEW

**Comment by dteleguin** [comment]
Doc PR: https://github.com/keycloak/keycloak-documentation/pull/838

File: themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties [hunk 1]
```diff
@@ -643,6 +643,10 @@ single-logout-service-url=Single Logout Service URL
 saml.single-logout-service-url.tooltip=The Url that must be used to send logout requests.
 nameid-policy-format=NameID Policy Format
 nameid-policy-format.tooltip=Specifies the URI reference corresponding to a name identifier format. Defaults to urn:oasis:names:tc:SAML:2.0:nameid-format:persistent.
+saml.principal-type=Principal
```
Comment (hmlnarik) [inline themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties]:
```suggestion
saml.principal-type=Principal Type
```

File: themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties [hunk 2]
```diff
@@ -643,6 +643,10 @@ single-logout-service-url=Single Logout Service URL
...
+saml.principal-type.tooltip=Specifies how Keycloak should identify and track external users. Default is using Subject NameID, though it is recommended to set up identifying attribute which is more reliable.
```
Comment (hmlnarik) [inline themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties]: No opinions please, just pure description.

```suggestion
saml.principal-type.tooltip=Way to identify and track external users from the assertion. Default is using Subject NameID, alternatively you can set up identifying attribute.
```

File: themes/src/main/resources-community/theme/base/admin/messages/admin-messages_ru.properties [hunk 3]
```diff
@@ -527,6 +527,10 @@ single-logout-service-url=Адреса сервиса единого выход
 saml.single-logout-service-url.tooltip=Url, который должен быть использован для отправленных запросов на выход.
 nameid-policy-format=Формат политики NameID
 nameid-policy-format.tooltip=Определяет ссылку URI, соответствующую формату идентификатора имени. По умолчанию urn:oasis:names:tc:SAML:2.0:nameid-format:persistent.
+saml.principal-type=Идентификация
```
Comment (hmlnarik) [inline themes/src/main/resources-community/theme/base/admin/messages/admin-messages_ru.properties]: Please address the comments from en version in this translation as well

File: services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java [hunk 4]
```diff
@@ -414,15 +416,24 @@ protected Response handleLoginResponse(String samlResponse, SAMLDocumentHolder h
                 SubjectType subject = assertion.getSubject();
                 SubjectType.STSubType subType = subject.getSubType();
                 NameIDType subjectNameID = (NameIDType) subType.getBaseID();
+                String principal = getPrincipal(assertion);
+
+                if (principal == null) {
+                    logger.error("no principal in assertion; expected: " + expectedPrincipalType());
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java]:
```suggestion
                    logger.errorf("no principal in assertion; expected: %s", expectedPrincipalType());
```

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thanks @dteleguin ! Overall looks good, few thing need to be addresses before merging though

**Review by hmlnarik — APPROVED** [review]

**Comment by hmlnarik** [comment]
Thanks @dteleguin !

**Pull request merged**